### PR TITLE
Input shape verification for convolutional models

### LINKED ISSUE #9: Input shape verification for convolutional models
### Proposal: 
Add a utility to validate and adjust input shapes for convolutional models.

### Description: 
Currently, the project does not provide a way to check whether an input shape is compatible with certain convolutional models. This can lead to runtime errors when users pass inputs that are incompatible with the model’s architecture. Examples are encoder-decoder models with skip connections that concatenate the decoded tensor and the skipped input; if during the encoding the pooling operation resulted in a non-integer size (which gets rounded), the concatenation can not happen. 

I propose adding utilities to:

Validate whether a given input shape is compatible with a specified model.
If the shape is incompatible, calculate and optionally apply the necessary padding to adjust the shape to the closest compatible size.

### Motivation 

This contribution will improve the usability of the models when used with various datasets with different shapes. 

### Proposed solution

Every model extending `ModelABC`  can optionally implement a function `validate_input_shape(self, input_shape: Size) -> Tuple[bool, Size]` which checks whether the current shape is fine, and, if it's not, it returns the closest shape that fits the model's architecture. 

A utility function `pad_batch(batch: torch.Tensor, new_shape: torch.Size, pad_value: float=0) -> torch.Tensor` can then be used to adjust your input batch's size. 

A proof of concept for `UNet`, `HalfUNet` and `CustomUNet` can be found [here](https://github.com/MicheleCattaneo/mfai/tree/check_input_shapes).

### Additional contributions

These utilities could be used inside the forward pass of the models; if there is a runtime error and the error is caused by the input shape, additional information is logged to the user to aid them with the process of padding the inputs. 

---

I would appreciate any feedback or suggestions on this approach, and I am willing to contribute and to proceed with a pull request once we’ve reviewed the proposal.

### PR DESCRIPTION
**Pull request by MicheleCattaneo**
### What does this PR do?
It introduces utilities to handle invalid input shapes for convolutional models. 

### Current progress
- [x] Functions to check the input shapes for `UNet`, `HalfUNet` and `CustomUNet` + tests.
- [x] Padding and un-padding functions to be used in conjunction with the aforementioned ones + tests.
- [x] Automatic padding for torch lightning modules exposed through their settings.

### Related Issue

closes #9 

This draft PR will be updated as the work progresses. Please share any feedback on the overall structure.

### CODE REVIEW

**Review by LBerth — COMMENTED** [review]
Hi !
Very good job and very nice first contribution.
I have only 2 minor comments concerning the unit test of the padding.
Léa

File: tests/test_utils.py
[the remark this answers is not in the document]
Comment (MicheleCattaneo) [reply tests/test_utils.py:1]: Good point, I will do it.

File: tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+def test_pad(dims):
+    # initial data with 16 batch elements, 3 channels and random input shape
+    tensor_shape = torch.Size([16,3]+[random.randint(60, 100) for _ in range(dims)])
+    data = torch.randn(*tensor_shape)
+    
+    # generate a new random input shape, larger than the original one
+    new_delta = [random.randint(5, 25) for _ in range(dims)]
+    new_shape = torch.Size([a + b for a, b in zip(tensor_shape[-dims:], new_delta)])
+    # pad the data 
+    padded_data = pad_batch(batch=data, new_shape=new_shape, pad_value=0)
```
[the remark this answers is not in the document]
Comment (MicheleCattaneo) [reply tests/test_utils.py:17]: I was afraid to not cover some edge cases  for odd/even sizes by keeping the sizes fixed. I am not sure if it's a good approach, but if there is an issue, it should come up with time (as tests are run multiple times).

**Comment by colon3ltocard** [comment]
Hi again @MicheleCattaneo ,

Sorry for the late reply we went thru a lot of changes before being able to integrate/discuss new features.

We discussed with the team and we have some suggestions to make for the autopad you propose:
* Separation Of Concern: auto-padding is only meaningful for a subset of the models so it should not "contaminate" outside the subset of models (the actual ModelABC subclasses) it is related to.

We propose that a model able to autopad should expose it thru extra settings, here is an example with HalfUnet:

```python
@dataclass_json
@dataclass(slots=True)
class HalfUNetSettings:
    num_filters: int = 64
    dilation: int = 1
    bias: bool = False
    use_ghost: bool = False
    last_activation: str = "Identity"
    absolute_pos_embed: bool = False
    autopad: bool = False
 ```
 
 Now the model can handle this in its forward method (this code could be factored in a common class or Mixin):
 
 ```python
 def forward(self, x):
      if self.settings.autopad:
           x = self.pad(x)
           y=self._forward(x)
           return self.unpad(y)
      else:
         return self._forward(x)
```
Note : i omitted the shape tracking on purpose for the brevity of this.

This way the autopad doesn't "pollute" the Segmentation/Regression lightning, we don't have to check if autopad is supported using dedicated code outside the model and we can also use it transparently in pure PyTorch loops.

**Comment by MicheleCattaneo** [comment]
Hi @colon3ltocard, 
I agree with what you suggest. Commit [337ef25](https://github.com/MicheleCattaneo/mfai/commit/337ef2529ceacb33020d7fc6c316affa5d14605e) implements a first variant of this idea. Let me know if you have any feedback!

File: mfai/torch/models/unet.py [hunk 2]
```diff
@@ -199,6 +209,29 @@ def _block(in_channels, features, name):
...
+
+
+        return new_shape == input_shape, new_shape
+    
+    @property
+    def _num_pool_layers(self):
+        # introspective, looks at the code of forword and 
+        # counts the number of max pool calls
+        source_code = inspect.getsource(self.forward)
+        return len(re.findall(r'max_pool\(', source_code))
```
Comment (colon3ltocard) [inline mfai/torch/models/unet.py:233]: maybe use a cachedproperty ? we probably don't want to inspect source code at each call.

File: mfai/torch/models/half_unet.py [hunk 3]
```diff
@@ -61,7 +64,7 @@ def forward(self, x):
         return self.relu(x)
 
 
-class HalfUNet(ModelABC, nn.Module):
+class HalfUNet(ModelABC, nn.Module, AutoPaddingModel):
```
Comment (colon3ltocard) [inline mfai/torch/models/half_unet.py:67]: For the sake of clarity I would put the two ABCs first:

```suggestion
class HalfUNet(ModelABC, AutoPaddingModel), nn.Module):
```

File: mfai/torch/models/unet.py [hunk 4]
```diff
@@ -207,9 +240,10 @@ class CustomUnetSettings:
     encoder_name: str = "resnet18"
     encoder_depth: int = 5
     encoder_weights: bool = True
+    autopad_enabled: bool = False
 
 
-class CustomUnet(ModelABC, nn.Module):
+class CustomUnet(ModelABC, nn.Module, AutoPaddingModel):
```
Comment (colon3ltocard) [inline mfai/torch/models/unet.py:246]:
```suggestion
class CustomUnet(ModelABC, AutoPaddingModel, nn.Module):
```

**Comment by colon3ltocard** [comment]
@MicheleCattaneo A few minor comments + some CI fails and then we can merge :+1:

File: mfai/torch/models/base.py [hunk 5]
```diff
@@ -91,3 +94,59 @@ def check_required_attributes(self):
...
+
+        Args:
+            data_tensor (torch.Tensor): The data tensor from which padding is to be removed. 
+            old_shape (torch.Size): The previous shape of the data tensor. It can either be 
+            [W,H] or [W,H,D] for 2D and 3D data respectively. old_shape is returned by self._maybe_padding.
+
+        Returns:
+            torch.Tensor: The data tensor with the padding removed, if possible.
+        """
+        if self.autopad_enabled and old_shape is not None:
```
Comment (colon3ltocard) [inline mfai/torch/models/base.py:150]:
```suggestion
        if self._settings.autopad_enabled and old_shape is not None:
```

File: mfai/torch/models/unet.py [hunk 6]
```diff
@@ -87,6 +93,7 @@ def __init__(
         self.in_channels = in_channels
         self.out_channels = out_channels
         self.input_shape = input_shape
+        self.autopad_enabled = settings.autopad_enabled
```
Comment (colon3ltocard) [inline mfai/torch/models/unet.py:96]:
```suggestion
```

**Review by colon3ltocard — CHANGES_REQUESTED** [review]
One last comment: directly reference the self._settings for the autopad_enabled bool

**Review by MicheleCattaneo — COMMENTED** [review]
Done 👍

**Review by colon3ltocard — APPROVED** [review]

**Review by colon3ltocard — APPROVED** [review]

**Review by LBerth — APPROVED** [review]
Hi ! Very good contribution, thank you Michele ! :)
